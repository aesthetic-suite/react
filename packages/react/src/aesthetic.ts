import { Aesthetic } from '@aesthetic/core';
import { createClientEngine } from '@aesthetic/style';

const aesthetic = new Aesthetic();

aesthetic.configureEngine(createClientEngine());

export default aesthetic;
