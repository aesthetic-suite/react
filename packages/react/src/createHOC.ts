/* eslint-disable @typescript-eslint/no-explicit-any */

import React from 'react';
import hoistNonReactStatics from 'hoist-non-react-statics';
import { WrapperComponent } from './types';

export default function createHOC<T extends React.FunctionComponent<any>>(
  name: string,
  WrappedComponent: React.ComponentType<any>,
  BaseComponent: T,
): T & WrapperComponent {
  const Component = BaseComponent as T & WrapperComponent;

  hoistNonReactStatics(Component, WrappedComponent);

  Component.displayName = `${name}(${WrappedComponent.displayName || WrappedComponent.name})`;
  Component.WrappedComponent = WrappedComponent;

  return Component;
}
