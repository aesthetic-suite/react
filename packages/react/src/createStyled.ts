import React from 'react';
import { createComponentStyles, LocalBlock, Utilities } from '@aesthetic/core';
import useStyles from './useStyles';

export default function createStyled<T extends keyof JSX.IntrinsicElements>(
  type: T,
  factory: (utilities: Utilities<LocalBlock>) => LocalBlock,
): React.ForwardRefExoticComponent<JSX.IntrinsicElements[T]> {
  if (__DEV__) {
    if (typeof factory !== 'function') {
      throw new TypeError(
        `Styled components require a style sheet factory function, found ${typeof factory}.`,
      );
    }
  }

  const styleSheet = createComponentStyles((utils) => ({
    element: factory(utils),
  }));

  const Component = React.forwardRef<unknown, JSX.IntrinsicElements[T]>((props, ref) => {
    const cx = useStyles(styleSheet);
    let className = cx('element');

    if (props.className) {
      className += ` ${props.className}`;
    }

    return React.createElement(type, {
      ...props,
      className,
      ref,
    });
  });

  Component.displayName = `styled(${type})`;

  return Component;
}
