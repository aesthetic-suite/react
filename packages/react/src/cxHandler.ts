import { ClassNameSheet, ClassNameSheetVariants, generateClassName } from '@aesthetic/core';
import { isObject, objectLoop } from '@aesthetic/utils';

const variantCache: Record<string, string[]> = {};

export default function cxHandler(keys: unknown[], classNames: ClassNameSheet<string>) {
  const names = keys as string[];
  let variantNames: string[] = [];

  if (isObject(names[0])) {
    const variants = (names.shift() as unknown) as ClassNameSheetVariants;
    const cacheKey = JSON.stringify(variants);
    const cache = variantCache[cacheKey];

    if (cache) {
      variantNames = cache;
    } else {
      objectLoop(variants, (value, type) => {
        variantNames.push(`${type}_${value}`);
      });

      variantCache[cacheKey] = variantNames;
    }
  }

  return generateClassName(names, variantNames, classNames);
}
