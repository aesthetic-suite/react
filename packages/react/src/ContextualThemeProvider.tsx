import React, { useState, useEffect } from 'react';
import aesthetic from './aesthetic';
import ThemeContext from './ThemeContext';
import useDirection from './useDirection';
import { ContextualThemeProviderProps } from './types';

/**
 * Provides a theme for a distinct portion of the React tree.
 */
export default function ContextualThemeProvider({
  children,
  name,
  wrapper,
}: ContextualThemeProviderProps) {
  const [className, setClassName] = useState('');
  const direction = useDirection();
  const theme = aesthetic.getTheme(name);

  // Render styles when theme/direction change
  useEffect(() => {
    setClassName(aesthetic.renderThemeStyles(theme, { direction }));
  }, [theme, direction]);

  return (
    <ThemeContext.Provider value={theme}>
      {React.cloneElement(wrapper, { className, 'data-theme': name }, children)}
    </ThemeContext.Provider>
  );
}
