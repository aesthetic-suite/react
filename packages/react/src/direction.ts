import { createDirectionHelpers } from '@aesthetic/core-react';
import { aesthetic } from './aesthetic';

export const {
	DirectionContext,
	DirectionProvider,
	useDirection,
	withDirection,
} = createDirectionHelpers(aesthetic, { wrapper: 'div' });
