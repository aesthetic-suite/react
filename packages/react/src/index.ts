/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { aesthetic } from './aesthetic';
import { createStyled } from './styles';

export * from './direction';
export * from './styles';
export * from './theme';
export * from './types';
export * from '@aesthetic/core';
export * from '@aesthetic/core-react';

export const {
	changeDirection,
	changeTheme,
	configure,
	createStyleSheet,
	createScopedStyleSheet,
	createThemeSheet,
	getActiveDirection,
	getActiveTheme,
	getEngine,
	getTheme,
	registerDefaultTheme,
	registerTheme,
	renderStyleSheet,
	renderFontFace,
	renderImport,
	renderKeyframes,
	renderThemeSheet,
	subscribe,
	unsubscribe,
} = aesthetic;

// Export aliases
export const style = createScopedStyleSheet;
export const styles = createStyleSheet;
export const styled = createStyled;
export const themeStyle = createThemeSheet;

// Do not use! This is for bundler pre-compilation usage!
// @private
export const internalAestheticRuntime = aesthetic;
