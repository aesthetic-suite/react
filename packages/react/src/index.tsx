/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import aesthetic from './aesthetic';
import ThemeContext from './ThemeContext';
import ThemeProvider from './ThemeProvider';
import ContextualThemeProvider from './ContextualThemeProvider';
import createStyled from './createStyled';
import useStyles from './useStyles';
import useTheme from './useTheme';
import withStyles from './withStyles';
import withTheme from './withTheme';

export * from '@aesthetic/core';
export * from './direction';
export * from './types';

export {
  ThemeContext,
  ThemeProvider,
  ContextualThemeProvider,
  createStyled,
  useStyles,
  useTheme,
  withStyles,
  withTheme,
};

export const {
  changeDirection,
  changeTheme,
  configure,
  createComponentStyles,
  createThemeStyles,
  generateClassName,
  getActiveDirection,
  getActiveTheme,
  getEngine,
  getTheme,
  registerDefaultTheme,
  registerTheme,
  renderComponentStyles,
  renderFontFace,
  renderImport,
  renderKeyframes,
  renderThemeStyles,
  subscribe,
  unsubscribe,
} = aesthetic;
