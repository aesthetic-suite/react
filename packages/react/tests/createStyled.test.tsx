/* eslint-disable react/jsx-no-literals */

import React from 'react';
import { render } from 'rut-dom';
import { getRenderedStyles, purgeStyles } from '@aesthetic/core/lib/testing';
import { createStyled } from '../src';
import { Wrapper } from './__mocks__/Button';
import { setupAestheticReact, teardownAestheticReact } from './helpers';

describe('createStyled()', () => {
  beforeEach(() => {
    setupAestheticReact();
  });

  afterEach(() => {
    purgeStyles();
    teardownAestheticReact();
  });

  it('creates and renders a button with defined styles', () => {
    const Button = createStyled('button', () => ({
      display: 'inline-flex',
      textAlign: 'center',
      padding: '1rem',
    }));

    const { debug } = render<{}>(<Button>Test</Button>, {
      wrapper: <Wrapper />,
    });

    expect(debug({ log: false })).toMatchSnapshot();
    expect(getRenderedStyles('standard')).toMatchSnapshot();
  });

  it('only renders styles once, even when component rerenders', () => {
    const spy = jest.fn(() => ({
      textDecoration: 'none',
    }));

    const Link = createStyled('a', spy);

    const { debug, update } = render<{}>(<Link href="/foo/bar">Test</Link>, {
      wrapper: <Wrapper theme="day" />,
    });

    update();
    update();
    update();

    expect(spy).toHaveBeenCalledTimes(1);
    expect(debug({ log: false })).toMatchSnapshot();
    expect(getRenderedStyles('standard')).toMatchSnapshot();
  });

  it('can pass custom props/attributes', () => {
    const Input = createStyled('input', {
      border: '1px solid black',
      padding: '1rem',
      ':focus': {
        outline: 'none',
      },
    });

    const { debug } = render<{}>(
      <Input disabled={false} type="text" placeholder="Search..." className="will-be-appended" />,
      {
        wrapper: <Wrapper />,
      },
    );

    expect(debug({ log: false })).toMatchSnapshot();
    expect(getRenderedStyles('standard')).toMatchSnapshot();
  });

  it('can use and define CSS variables', () => {
    const Code = createStyled('code', (css) => ({
      fontSize: css.var('text-sm-size'),
      lineHeight: css.token('text-sm-line-height') as number,
      '@variables': {
        elementLevelVar: 'nice',
      },
    }));

    const { debug } = render<{}>(<Code />, {
      wrapper: <Wrapper />,
    });

    expect(debug({ log: false })).toMatchSnapshot();
    expect(getRenderedStyles('standard')).toMatchSnapshot();
  });

  it('supports variants', () => {
    interface AlertProps {
      palette?: 'success' | 'failure' | 'warning';
    }

    const Alert = createStyled<'div', AlertProps>('div', (css) => ({
      background: css.var('palette-neutral-bg-base'),
      '@variants': {
        palette: {
          success: {
            background: css.var('palette-success-bg-base'),
          },
          failure: {
            background: css.var('palette-failure-bg-base'),
          },
          warning: {
            background: css.var('palette-warning-bg-base'),
          },
        },
      },
    }));

    const { debug, update } = render<AlertProps>(
      <Alert palette="success">
        <div>
          <h1>Title</h1>And other content!
        </div>
      </Alert>,
      {
        wrapper: <Wrapper />,
      },
    );

    expect(debug({ log: false })).toMatchSnapshot();
    expect(getRenderedStyles('standard')).toMatchSnapshot();

    update({ palette: 'failure' });

    expect(debug({ log: false })).toMatchSnapshot();

    update({ palette: undefined });

    expect(debug({ log: false })).toMatchSnapshot();
  });

  it('supports styled component composition', () => {
    const Button = createStyled('button', () => ({
      display: 'inline-flex',
      textAlign: 'center',
      padding: '1rem',
    }));

    const BlockButton = createStyled(Button, {
      display: 'flex',
      width: '100%',
    });

    const LargeBlockButton = createStyled(BlockButton, {
      padding: '2rem',
      fontSize: 18,
    });

    const { debug } = render<{}>(
      <>
        <Button type="button">Normal</Button>
        <BlockButton type="submit">Block</BlockButton>
        <LargeBlockButton disabled>Large</LargeBlockButton>
      </>,
      {
        wrapper: <Wrapper />,
      },
    );

    expect(debug({ log: false })).toMatchSnapshot();
    expect(getRenderedStyles('standard')).toMatchSnapshot();
  });

  it('can access the ref from multiple layers of composition', () => {
    const Leaf = createStyled('main', {});
    const Branch = createStyled(Leaf, {});
    const Trunk = createStyled(Branch, {});
    const Root = createStyled(Trunk, {});

    const svg = document.createElement('main');
    const spy = jest.fn();
    const { debug } = render<{}>(<Root ref={spy} />, {
      mockRef: () => svg,
      wrapper: <Wrapper />,
    });

    expect(spy).toHaveBeenCalledWith(svg);
    expect(debug({ log: false })).toMatchSnapshot();
  });
});
