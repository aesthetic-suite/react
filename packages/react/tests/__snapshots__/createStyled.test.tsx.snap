// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createStyled() can access the ref from multiple layers of composition 1`] = `
"<Wrapper>
  <DirectionProvider direction=\\"ltr\\" wrapper={<div />}>
    <div dir=\\"ltr\\">
      <ThemeProvider name=\\"twilight\\">
        <ForwardRef>
          <ForwardRef className=\\"\\">
            <ForwardRef className=\\"\\">
              <ForwardRef className=\\"\\">
                <main ref={mockConstructor()} className=\\"\\" />
              </ForwardRef>
            </ForwardRef>
          </ForwardRef>
        </ForwardRef>
      </ThemeProvider>
    </div>
  </DirectionProvider>
</Wrapper>"
`;

exports[`createStyled() can pass custom props/attributes 1`] = `
"<Wrapper>
  <DirectionProvider direction=\\"ltr\\" wrapper={<div />}>
    <div dir=\\"ltr\\">
      <ThemeProvider name=\\"twilight\\">
        <ForwardRef
          className=\\"will-be-appended\\"
          placeholder=\\"Search...\\"
          type=\\"text\\"
        >
          <input
            className=\\"a b c will-be-appended\\"
            placeholder=\\"Search...\\"
            type=\\"text\\"
          />
        </ForwardRef>
      </ThemeProvider>
    </div>
  </DirectionProvider>
</Wrapper>"
`;

exports[`createStyled() can pass custom props/attributes 2`] = `".a {border: 1px solid black;}.b {padding: 1rem;}.c:focus {outline: none;}"`;

exports[`createStyled() can use and define CSS variables 1`] = `
"<Wrapper>
  <DirectionProvider direction=\\"ltr\\" wrapper={<div />}>
    <div dir=\\"ltr\\">
      <ThemeProvider name=\\"twilight\\">
        <ForwardRef>
          <code className=\\"a b\\" />
        </ForwardRef>
      </ThemeProvider>
    </div>
  </DirectionProvider>
</Wrapper>"
`;

exports[`createStyled() can use and define CSS variables 2`] = `".a {font-size: var(--text-sm-size);}.b {line-height: 1.25;}"`;

exports[`createStyled() creates and renders a button with defined styles 1`] = `
"<Wrapper>
  <DirectionProvider direction=\\"ltr\\" wrapper={<div />}>
    <div dir=\\"ltr\\">
      <ThemeProvider name=\\"twilight\\">
        <ForwardRef>
          <button className=\\"a b c\\">
            Test
          </button>
        </ForwardRef>
      </ThemeProvider>
    </div>
  </DirectionProvider>
</Wrapper>"
`;

exports[`createStyled() creates and renders a button with defined styles 2`] = `".a {display: inline-flex;}.b {text-align: center;}.c {padding: 1rem;}"`;

exports[`createStyled() only renders styles once, even when component rerenders 1`] = `
"<Wrapper theme=\\"day\\">
  <DirectionProvider direction=\\"ltr\\" wrapper={<div />}>
    <div dir=\\"ltr\\">
      <ThemeProvider name=\\"day\\">
        <ForwardRef href=\\"/foo/bar\\">
          <a className=\\"a\\" href=\\"/foo/bar\\">
            Test
          </a>
        </ForwardRef>
      </ThemeProvider>
    </div>
  </DirectionProvider>
</Wrapper>"
`;

exports[`createStyled() only renders styles once, even when component rerenders 2`] = `".a {text-decoration: none;}"`;

exports[`createStyled() supports styled component composition 1`] = `
"<Wrapper>
  <DirectionProvider direction=\\"ltr\\" wrapper={<div />}>
    <div dir=\\"ltr\\">
      <ThemeProvider name=\\"twilight\\">
        <ForwardRef type=\\"button\\">
          <button className=\\"a b c\\" type=\\"button\\">
            Normal
          </button>
        </ForwardRef>
        <ForwardRef type=\\"submit\\">
          <ForwardRef className=\\"d e\\" type=\\"submit\\">
            <button className=\\"a b c d e\\" type=\\"submit\\">
              Block
            </button>
          </ForwardRef>
        </ForwardRef>
        <ForwardRef disabled>
          <ForwardRef disabled className=\\"f g\\">
            <ForwardRef disabled className=\\"d e f g\\">
              <button disabled className=\\"a b c d e f g\\">
                Large
              </button>
            </ForwardRef>
          </ForwardRef>
        </ForwardRef>
      </ThemeProvider>
    </div>
  </DirectionProvider>
</Wrapper>"
`;

exports[`createStyled() supports styled component composition 2`] = `".a {display: inline-flex;}.b {text-align: center;}.c {padding: 1rem;}.d {display: flex;}.e {width: 100%;}.f {padding: 2rem;}.g {font-size: 18px;}"`;

exports[`createStyled() supports variants 1`] = `
"<Wrapper>
  <DirectionProvider direction=\\"ltr\\" wrapper={<div />}>
    <div dir=\\"ltr\\">
      <ThemeProvider name=\\"twilight\\">
        <ForwardRef palette=\\"success\\">
          <div className=\\"a b\\" palette=\\"success\\">
            <div>
              <h1>Title</h1>
              And other content!
            </div>
          </div>
        </ForwardRef>
      </ThemeProvider>
    </div>
  </DirectionProvider>
</Wrapper>"
`;

exports[`createStyled() supports variants 2`] = `".a {background: var(--palette-neutral-bg-base);}.b {background: var(--palette-success-bg-base);}.c {background: var(--palette-failure-bg-base);}.d {background: var(--palette-warning-bg-base);}"`;

exports[`createStyled() supports variants 3`] = `
"<Wrapper>
  <DirectionProvider direction=\\"ltr\\" wrapper={<div />}>
    <div dir=\\"ltr\\">
      <ThemeProvider name=\\"twilight\\">
        <ForwardRef palette=\\"failure\\">
          <div className=\\"a c\\" palette=\\"failure\\">
            <div>
              <h1>Title</h1>
              And other content!
            </div>
          </div>
        </ForwardRef>
      </ThemeProvider>
    </div>
  </DirectionProvider>
</Wrapper>"
`;

exports[`createStyled() supports variants 4`] = `
"<Wrapper>
  <DirectionProvider direction=\\"ltr\\" wrapper={<div />}>
    <div dir=\\"ltr\\">
      <ThemeProvider name=\\"twilight\\">
        <ForwardRef>
          <div className=\\"a\\">
            <div>
              <h1>Title</h1>
              And other content!
            </div>
          </div>
        </ForwardRef>
      </ThemeProvider>
    </div>
  </DirectionProvider>
</Wrapper>"
`;
